example_ams/ams/preprocess_binary_images.py

```diff
@@ -0,0 +1,82 @@
+#
+# Copyright (c) 2018 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from typing import Tuple
+
+import tensorflow as tf
+import numpy as np
+
+
+def preprocess_binary_image(image: bytes, channels: int = None,
```

Comment:
Let's not propagate the underlying exceptions out of this function. It's supposed to provide an abstraction for the pre-processing functionality however currently it leaks the details of the underlying implementation by propagating the exceptions from the specific library used under the hood - if that library changes, the surrounding code that uses that function will also need to change. Instead let's identify the possible failure modes for the preprocessing functionality, define them as possible status codes in the contract of that function and then make sure that the errors reported by the underlying library are correctly interpreted and reported.

Reply:
Only one exception may be raised by this function, as it is written in the docstring below: ValueError. Therefore, no library specific exceptions have to be expected by the caller, and potential underlying library changes will have no effect on the code that is using this function. There is a clear contract between this function and its callers - caller has to only expect ValueError - which will contain all the information about the actual cause of error, so no information will be lost. Moreover, ValueError is widely used for such purposes and it is a recommended way to indicate an error which was probably caused by invalid value passed to the function - in this case, we expect that errors will be caused by invalid value of passed image.

If you are not familiar with the concept of exception chaining, see: https://www.python.org/dev/peps/pep-3134

Reply:
Thanks, I'm familiar with the concept of exception chaining. And the concern here **is not possible loss of information**. The point here is that ValueError semantics (specifically the circumstances under which the exception can be thrown) will rely on the specific underlying implementation of the libraries used and anyone using that function who wants to communicate meaningful errors to their users will need to refer to that underlying implementation to understand which situations may result in exceptions and how to meaningfully tell one category of problems from another (data error, format error, parameter error, etc.). This is shifting the responsibility of understanding the possible failure modes in that function from its author to the users of that function which is not good encapsulation.

Reply:
I don't agree that providing caller with full information about the cause of failure is good encapsulation - it is widely considered as good practice to provide caller with possibility to look at full stack trace in case of error. Although, it is a good point that more granular information about the cause of error would be beneficial. So, my proposition is to handle it in following manner:
```python
try:
        decoded_image = tf.io.decode_image(image, channels=channels, dtype=dtype)
except Exception as e:
        raise ValueError('Failed to decode provided binary image.') from e

try:
        if target_size:
            height, width = target_size
            decoded_image = tf.image.resize_with_crop_or_pad(decoded_image, height, width)
except Exception as e:
      raise ValueError('Failed to resize provided binary image') from e
 (...)
```
That way caller would get more detailed information about the cause of error, without the need to create a new abstraction around returning statuses instead of raising proper exceptions. What do you think about this?

Reply:
By the way, please look at the documentation of https://www.tensorflow.org/api_docs/python/tf/io/decode_image - it defines exactly the same contract as in my solution - `ValueError` is the only exception that should be expected from this function.